File: find-peak-element.cpp
//Runtime: 6 ms
class Solution {
public:
	int binary(vector<int>&A, int lo, int hi)
	{
		if (lo == hi)
			return lo;

		int mid = lo + (hi - lo) / 2;
		if (A[mid] > A[mid + 1])
			return binary(A, lo, mid);

		return binary(A, mid + 1, hi);
	}

	int findPeakElement(vector<int>& nums) {
		return binary(nums, 0, nums.size() - 1);
	}
};

File: find-peak-element.scala
//Runtime: 344 ms
object Solution {
    def findPeakElement(nums: Array[Int]): Int = {
        def solve(nums: Array[Int], lo: Int, hi: Int): Int = {
            val mid = lo + (hi - lo)/2;
            if(lo == hi) lo
            else if(nums(mid) > nums(mid+1)) solve(nums, lo, mid)
            else solve(nums, mid+1, hi)
        }
        solve(nums, 0, nums.size - 1)
    }
}